Source/ablastr/particles/DepositCharge.H

```diff
@@ -28,51 +30,78 @@ namespace particles {
- * \param n_rz_azimuthal_modes number of azimuthal modes in use, irrelevant outside RZ geometry.
- * \param load_balance_costs_update_algo selected method for updating load balance costs.
- * \param do_device_synchronize call amrex::Gpu::synchronize() for tiny profiler regions
+               particles deposit current. If nullptr, costs are not updated. (default: nullptr)
+ * \param load_balance_costs_update_algo selected method for updating load balance costs (default: 0)
+ * \param do_device_synchronize call amrex::Gpu::synchronize() for tiny profiler regions (default: true)
  */
 template< typename T_PC >
 static void
 deposit_charge (typename T_PC::ParIterType& pti,
                 typename T_PC::RealVector const& wp,
-                const int * const ion_lev,
-                amrex::MultiFab* rho, const int icomp, const int nc,
-                const long offset, const long np_to_depose,
-                amrex::FArrayBox& local_rho, const int lev, const int depos_lev,
-                const amrex::Real charge, const int nox, const int noy, const int noz,
-                const amrex::IntVect& ng_rho, const std::array<amrex::Real,3>& dx,
-                const std::array<amrex::Real, 3>& xyzmin,
-                const amrex::IntVect& ref_ratio,
-                amrex::Real* cost, const int n_rz_azimuthal_modes,
-                const long load_balance_costs_update_algo,
-                const bool do_device_synchronize)
+                amrex::Real const charge,
+                int const * const ion_lev,
+                amrex::MultiFab* rho,
+                amrex::FArrayBox& local_rho,
+                int const particle_shape,
+                std::array<amrex::Real, 3> const & dx,
+                std::array<amrex::Real, 3> const & xyzmin,
+                int const n_rz_azimuthal_modes = 0,
+                std::optional<amrex::IntVect> num_rho_deposition_guards = std::nullopt,
+                std::optional<int> depos_lev = std::nullopt,
+                std::optional<amrex::IntVect> rel_ref_ratio = std::nullopt,
+                long const offset = 0,
+                std::optional<long> np_to_depose = std::nullopt,
+                int const icomp = 0, int const nc = 1,
+                amrex::Real * const AMREX_RESTRICT cost = nullptr,
+                long const load_balance_costs_update_algo = 0,
+                bool const do_device_synchronize = true)
 {
-    AMREX_ALWAYS_ASSERT_WITH_MESSAGE((depos_lev==(lev-1)) ||
-                                     (depos_lev==(lev  )),
-                                     "Deposition buffers only work for lev-1");
+    // deposition guards
+    amrex::IntVect ng_rho = rho->nGrowVect();
+    if (num_rho_deposition_guards.has_value())
+        ng_rho = num_rho_deposition_guards.value();
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(ng_rho <= rho->nGrowVect(),
+                                     "num_rho_deposition_guards are larger than allocated!");
```

Comment:
Same question here: I wonder if we can check this only once initially? The case that is not trivial is the one where `ng_rho` is actually set to `num_rho_deposition_guards.value()`, and we know that this boils down to making sure that `ng_depos_rho <= ng_alloc_rho` when these are set in the `guardCellManager`. Do you think that checking this only once initially would not be robust enough?

Reply:
This is a separate API call and will be called by other codes, such as ImpactX.

Thus, we need to check these arguments every time they come in.
These checks are super cheap and not in a hot loop, thus they belong exactly here where we validate the user inputs.

Reply:
Same, sounds good to me then.